json-schema-generator/src/main/java/io/micronaut/jsonschema/generator/SourceGenerator.java

```diff
@@ -0,0 +1,638 @@
+        discriminatorProperty = discriminator.getPropertyName();
+
+        List<AnnotationDef> subTypeList = discriminator.getMapping().entrySet()
+            .stream()
+            .map(entry -> AnnotationDef
+                .builder(JsonSubTypes.Type.class)
+                .addMember("value", context.getDefinitionType(inputFileName + entry.getValue()))
+                .addMember("name", entry.getKey())
+                .build())
+            .toList();
+        AnnotationDef jsonSubTypes = AnnotationDef.builder(JsonSubTypes.class).addMember("value", subTypeList).build();
+        objectBuilder.addAnnotation(jsonSubTypes);
+    }
+
+    private String getJavadoc(String description) {
+        if (description.isBlank()) {
+            return "";
+        }
+        return description
+            .replaceAll("<", "&lt;")
+            .replaceAll(">", "&gt;")
+            .replaceAll("&", "&amp;")
+            .replaceAll("'", "&apos;")
+            .replaceAll("\"", "&quot;")
+            .replaceAll("\n", "<br>")
+            .trim();
+    }
+
+    private TypeDef getEnumType(ObjectDefBuilder objectBuilder, String propertyName, Schema schema) {
+        EnumDef enumDef = buildEnum(schema, capitalize(propertyName));
+        objectBuilder.addInnerType(enumDef);
+        return enumDef.asTypeDef();
+    }
+
+    private TypeDef getListTypeDef(ObjectDefBuilder objectBuilder, String propertyName, Schema schema) {
+        Schema items = schema.getItems() != null ? schema.getItems() : schema.getContains();
+        if (items == null) {
+            return TypeDef.OBJECT;
+        }
+
+        TypeDef propertyType = getTypeDefFromJson(items, context);
+        if (propertyType.equals(TypeDef.of(List.class))) {
+            propertyType = getListTypeDef(objectBuilder, propertyName, items);
+        } else if (items.isEnum()) {
+            propertyType = getEnumType(objectBuilder, propertyName, items);
+        } else if (propertyType instanceof TypeDef.Primitive primitive) {
+            propertyType = primitive.wrapperType();
+        }
+
+        if (propertyType.equals(TypeDef.OBJECT) && items.hasProperties()) {
+            propertyType = buildInnerType(objectBuilder, propertyName, items);
+        }
+
+        var annotations = AnnotationsAggregator.getAnnotations(items, propertyType);
+        return TypeDef.parameterized(
+            (schema.isUniqueItems() != null && schema.isUniqueItems()) ? Set.class : List.class,
+            propertyType.annotated(annotations));
+    }
+
+    private TypeDef buildInnerType(ObjectDefBuilder objectBuilder, String propertyName, Schema schema) {
+        // inner type
+        ObjectDef builder;
+        if (schema.getProperties().size() > 255 || schema.hasAdditionalProperties() || schema.hasConstValue()) {
+            builder = buildClass(schema, capitalize(propertyName));
+        } else {
+            builder = buildRecord(schema, capitalize(propertyName));
+        }
```

Comment:
Can this be an enum?

Reply:
Enum is checked before this point, so no.